lib/AST/ASTVerifier.cpp

```diff
@@ -2367,6 +2367,20 @@ class Verifier : public ASTWalker {
         }
       }
 
+      if (VD->isFinal()) {
+        if (!VD->getAttrs().hasAttribute<FinalAttr>()) {
+          Out << "decl cannot be 'final' without FinalAttr\n";
+          VD->dump(Out);
+          abort();
+        }
+      } else {
+        if (VD->getAttrs().hasAttribute<FinalAttr>()) {
+          Out << "decl cannot have FinalAttr and not be 'final'\n";
+          VD->dump(Out);
+          abort();
+        }
+      }
```

Comment:
Took me a few seconds to get the double negative "cannot ... not be" and realize what exactly we're trying to do. I think combining the checks would be easier to understand:

```
if (VD->isFinal() != VD->getAttrs().hasAttribute<FinalAttr>()) {
  Out << "decl should be final iff it has FinalAttr, but isFinal() = " << VD->isFinal() 
      << " and hasAttribute<FinalAttr> = " << VD->getAttrs().hasAttribute<FinalAttr>()
      << '\n';
  [..]
}
```

Reply:
I considered that, but we haven't been using that kind of complicated message in the verifier. I can't think of a good reason why *not* though, so I'll change it after the tests run.